ports/stm32/ulpi.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2016 STMicroelectronics. All rights reserved.
+ *
+ * This software component is licensed by ST under BSD 3-Clause license,
+ * the "License"; You may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *                        opensource.org/licenses/BSD-3-Clause
+ *
+ * USB3320 ULPI functions ported from stm32f7xx_lp_modes.c
+ */
+#include <stdint.h>
+#include "py/mphal.h"
+#include "pin_static_af.h"
+#include "ulpi.h"
+
+#if (MICROPY_HW_USB_HS_ULPI3320 == 1)
+#include STM32_HAL_H
+
+#define USBULPI_PHYCR       ((uint32_t)(0x40040000 + 0x034))
+#define USBULPI_D07         ((uint32_t)0x000000FF)
+#define USBULPI_New         ((uint32_t)0x02000000)
+#define USBULPI_RW          ((uint32_t)0x00400000)
+#define USBULPI_S_BUSY      ((uint32_t)0x04000000)
+#define USBULPI_S_DONE      ((uint32_t)0x08000000)
+
+#define USB_OTG_READ_REG32(reg)  (*(__IO uint32_t *)(reg))
+#define USB_OTG_WRITE_REG32(reg,value) (*(__IO uint32_t *)(reg) = (value))
+
+extern void __fatal_error(const char *msg);
+
+/**
+  * @brief  Read CR value
+  * @param  Addr the Address of the ULPI Register
+  * @retval Returns value of PHY CR register
+  */
+static uint32_t USB_ULPI_Read(uint32_t Addr) {
+    __IO uint32_t val = 0;
+    __IO uint32_t timeout = 100; /* Can be tuned based on the Clock or etc... */
+
+    USB_OTG_WRITE_REG32(USBULPI_PHYCR, USBULPI_New | (Addr << 16));
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    while (((val & USBULPI_S_DONE) == 0) && (timeout--)) {
+        val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    }
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    return val & 0x000000ff;
+}
+
+/**
+  * @brief  Write CR value
+  * @param  Addr the Address of the ULPI Register
+  * @param  Data Data to write
+  * @retval Returns value of PHY CR register
+  */
+static uint32_t USB_ULPI_Write(uint32_t Addr, uint32_t Data) {
+    __IO uint32_t val;
+    __IO uint32_t timeout = 10;   /* Can be tuned based on the Clock or etc... */
+
+    USB_OTG_WRITE_REG32(USBULPI_PHYCR, USBULPI_New | USBULPI_RW | (Addr << 16) | (Data & 0x000000ff));
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    while (((val & USBULPI_S_DONE) == 0) && (timeout--)) {
+        val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    }
+
+    val = USB_OTG_READ_REG32(USBULPI_PHYCR);
+    return 0;
+}
+
+/**
+  * @brief  This function configures the USB PHY to enter the low power mode
+  * @param  None
+  * @retval None
+  */
+void ulpi_enter_low_power(void) {
+    __IO uint32_t regval = 0;
+
+    /* disable ULPI_CLK by accessing ULPI_PHY */
+    /* read Vendor ID : (Low, High) 0x24,0x04 for USB3300 */
+    regval = USB_ULPI_Read(0x00);
+    if (regval != 0x24) {
+        __fatal_error("ULPI Error 0x00 != 0x24");
```

Comment:
Does it make sense to call fatal error?  Maybe better to return an -MP_Exxx error code and let the caller deal with error cases?

Reply:
It makes it harder to know which read/write failed, but this code has been stable for a while so I guess it's fine.